File: 0202 Happy Number/solution.cpp
#include "iostream"
#include "unordered_set"
using namespace std;

class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> st;
        while (true)
        {
            int t = 0;
            while (n)
            {
                t += (n % 10) * (n % 10);
                n /= 10;
            }

            if (t == 1)
                return true;
            else if (st.find(t) != st.end())
                return false;
            else
                st.insert(t);
            n = t;
        }
    }
};

int main(){

    cout << Solution().isHappy(19) << endl;
    cout << Solution().isHappy(22) << endl;
    return 0;
}

File: 0202 Happy Number/solution.py
#! python3

class Solution:
    def isHappy(self, n):
        slow = self.helper(n)
        fast = self.helper(slow)
        
        while slow != fast :
            slow = self.helper(slow)
            fast = self.helper(self.helper(fast))
        
        if slow == 1:
            return True
        else:
            return False

    def helper(self, n):
        r = 0
        while n:
            r += (n % 10) ** 2
            n //= 10
        
        return r
    
if __name__ == "__main__":
    print(Solution().isHappy(19))
    print(Solution().isHappy(22))